### PR TITLE
feat: Add .chezmoiroot to allow source state to be subdir of source dir

### LINKED ISSUE #1519: Support `.chezmoiroot` file to redirect `init`'s source dir to subfolder
## Is your feature request related to a problem? Please describe.

This is a follow-up of https://github.com/twpayne/chezmoi/issues/891#issuecomment-933742656. The problem of having to run

```
sh <(curl -fsSL https://git.io/chezmoi) -- init --apply --working-tree=~/.local/share/chezmoi --source=~/.local/share/chezmoi/home felipecrs
```

Is that it's a long command, difficult to remember, and breaks the fact that `chezmoi init` would not work by default on all the repositories.

## Describe the solution you'd like

A new file named `.chezmoiroot` (the name is debatable), which I would put in the root of my dotfiles repository and would have contents like so:

```
$ cat .chezmoiroot
home
```

It would tell chezmoi that, when initializing the repository, it should adjust the `sourceDir` to be the `home` subfolder automatically.

This would allow me to remove some workarounds that I have to circumvent this, such as [this](https://github.com/felipecrs/dotfiles/commit/9cafef4d988ce652c0fcf9ad1ea9fc24b3f58c46#diff-1e142e6277b12b7e1110478a24caee8f006a9349e86970c890203d6266209463R35).

I think this is the only thing missing to declare first-class support for subfolders in the dotfiles directory as the source dir. With it in place, the following would work:

```console
sh <(curl -fsSL https://git.io/chezmoi) -- init --apply felipecrs
```

## Describe alternatives you've considered

Described earlier.

## Additional context

I'm quite happy with my dotfiles being in a subdirectory. Previously, I oftenly created wrong files in my home dir because I forgot to add to .chezmoiignore, now I don't have to remember this step anymore.

### PR DESCRIPTION
**Pull request by twpayne**
Fixes #1519.

### CODE REVIEW

File: docs/HOWTO.md [hunk 1]
```diff
@@ -1601,6 +1602,36 @@ sudo apt install -y vim-gtk
...
 
+### Use a subdirectory of your dotfiles repo as the root of the source state
+
+By default, chezmoi uses the root of your dotfiles repo as the root of the
+source state. If your source state contains a lot of entries in its root, then
+your dotfiles repo will in turn contain a lot of entries in its root.
+
+You can instead specify that chezmoi should read the source state from a
+subdirectory of the source directory instead by creating a file called
+`.chezmoiroot` containing the relative path the source state. This reduces the
```
Comment (felipecrs) [inline docs/HOWTO.md:1613]:
```suggestion
`.chezmoiroot` containing the relative path to the source state. This reduces the
```

File: docs/HOWTO.md [hunk 2]
```diff
@@ -1601,6 +1602,36 @@ sudo apt install -y vim-gtk
 
 ---
 
+### Use a subdirectory of your dotfiles repo as the root of the source state
+
+By default, chezmoi uses the root of your dotfiles repo as the root of the
+source state. If your source state contains a lot of entries in its root, then
+your dotfiles repo will in turn contain a lot of entries in its root.
```
Comment (felipecrs) [inline docs/HOWTO.md:1609]:
```suggestion
your target folder (usually your home) will in turn be filled with a lot of entries in its root as well.
```

File: docs/HOWTO.md [hunk 3]
```diff
@@ -1601,6 +1602,36 @@ sudo apt install -y vim-gtk
...
+
+You can instead specify that chezmoi should read the source state from a
+subdirectory of the source directory instead by creating a file called
+`.chezmoiroot` containing the relative path the source state. This reduces the
+number of files in the root of your dotfiles repo and reduces the number of
```
Comment (felipecrs) [inline docs/HOWTO.md:1614]:
```suggestion
number of files in the root of your target folder and reduces the number of
```

```suggestion
number of
```

File: docs/HOWTO.md [hunk 4]
```diff
@@ -1601,6 +1602,36 @@ sudo apt install -y vim-gtk
 
 ---
 
+### Use a subdirectory of your dotfiles repo as the root of the source state
+
+By default, chezmoi uses the root of your dotfiles repo as the root of the
+source state. If your source state contains a lot of entries in its root, then
+your dotfiles repo will in turn contain a lot of entries in its root.
+
```
Comment (felipecrs) [inline docs/HOWTO.md:1610]: It's probably a good idea to first explain the first solution, which would be to keep the `.chezmoiignore` updated. And also to mention why or where it would be a better idea to use `.chezmoiroot`. E.g.: with `.chezmoiroot`, if you add a new file to the root of the repository, you don't need to remember to manually add it to the `.chezmoiignore`.

File: docs/HOWTO.md [hunk 5]
```diff
@@ -1601,6 +1602,36 @@ sudo apt install -y vim-gtk
...
+You can instead specify that chezmoi should read the source state from a
+subdirectory of the source directory instead by creating a file called
+`.chezmoiroot` containing the relative path the source state. This reduces the
+number of files in the root of your dotfiles repo and reduces the number of
+files that you need to explicitly exclude with `.chezmoiignore`.
```
Comment (felipecrs) [inline docs/HOWTO.md:1615]: But if you add that paragraph about explaining chezmoignore's approach, this last sentence may not be needed anymore.

File: docs/HOWTO.md [hunk 6]
```diff
@@ -1601,6 +1602,36 @@ sudo apt install -y vim-gtk
...
+
+For example, if `.chezmoiroot` contains:
+
+```
+home
+```
+
+Then chezmoi will read the source state from the `home` subdirectory of your
+source directory, for example the desired state of `~/.gitconfig` will be read
+from `~/.local/share/chezmoi/home/dot_gitconfig`.
```
Comment (felipecrs) [inline docs/HOWTO.md:1625]: Just to make it even more clear.

```suggestion
from `~/.local/share/chezmoi/home/dot_gitconfig` (instead of `~/.local/share/chezmoi/dot_gitconfig`).
```

File: internal/cmd/config.go [hunk 7]
```diff
@@ -1278,6 +1278,16 @@ func (c *Config) newRootCmd() (*cobra.Command, error) {
...
 	sourceStateLogger := c.logger.With().Str(logComponentKey, logComponentValueSourceState).Logger()
+
+	sourceDirAbsPath := c.SourceDirAbsPath
+	switch data, err := c.sourceSystem.ReadFile(sourceDirAbsPath.JoinString(chezmoi.RootName)); {
+	case errors.Is(err, fs.ErrNotExist):
+	case err != nil:
+		return nil, err
+	default:
+		sourceDirAbsPath = sourceDirAbsPath.JoinString(string(bytes.TrimSpace(data)))
+	}
```
Comment (felipecrs) [inline internal/cmd/config.go:1289]: I'm wondering... shouldn't `workingTree` be used as base here rather than `sourceDir` itself? I don't know. Did you think about it? The answer is probably no, but it's better to make sure we thought on everything.

File: internal/cmd/testdata/scripts/root.txt [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+cmp $CHEZMOISOURCEDIR/home/symlink_dot_symlink golden/symlink_dot_symlink
+
+-- golden/.file --
+# contents of .file
+-- golden/symlink_dot_symlink --
+.file
+-- home/user/.local/share/chezmoi/.chezmoiroot --
+home
+-- home/user/.local/share/chezmoi/home/dot_file --
+# contents of .file
```
Comment (felipecrs) [inline internal/cmd/testdata/scripts/root.txt:17]: perhaps it worth it to add a test for `init` with a config template as well?

**Review by felipecrs — COMMENTED** [review]
It's probably good to go. None of my suggestions are actually required, they are just to fine-tune.

File: internal/cmd/testdata/scripts/root.txt (hunk 8 above)
Comment (twpayne) [reply internal/cmd/testdata/scripts/root.txt:17]: Done (in my working copy, not pushed yet).

File: docs/HOWTO.md (hunk 4 above)
Comment (twpayne) [reply docs/HOWTO.md:1610]: Done (in working copy).

File: docs/HOWTO.md (hunk 5 above)
Comment (twpayne) [reply docs/HOWTO.md:1615]: Done (in working copy).

File: internal/cmd/config.go (hunk 7 above)
Comment (twpayne) [reply internal/cmd/config.go:1289]: This is an extremely good point, and I think you're right. I've updated the code and the docs to interpret `.chezmoiroot` as relative to the working tree instead of the source directory.

With these three directories in play (source dir, working tree, and chezmoi root) then it's likely that I've missed some corner cases. Currently the tests pass, but the tests are not exhaustive. If there are some tests that demonstrate problems with the working dir approach then I would be very happy to add them :)

**Comment by twpayne** [comment]
PR updated, please take another look.

File: internal/cmd/config.go (hunk 7 above)
Comment (felipecrs) [reply internal/cmd/config.go]: My only remaining concern is: if the user set the `sourceDir` in the configuration file, it will take precedence. Right?

Comment (twpayne) [reply internal/cmd/config.go]: On further thought, I think `.chezmoiroot` has to be read from the source directory, otherwise the `--source` option will be ignored if `.chezmoiroot` is present in the working tree. Unfortunately with the flag parsing library that chezmoi uses (github.com/spf13/cobra) I don't think it's possible to distinguish between "the user did not specify the flag so it's set to the default value" and "the user did specify the flag and passed the default value".

Comment (felipecrs) [reply internal/cmd/config.go]: I agree.

> I don't think it's possible to distinguish between "the user did not specify the flag so it's set to the default value" and "the user did specify the flag and passed the default value".

I suppose there will not be a need for that if chezmoi reads chezmoiroot from source directory rather than from working tree.
